### ISSUE TITLE
Can LLAMA 3 be trained on indices?

### ISSUE THREAD

**Issue by wiiiktor** [teticio/llama-squad#7]
Can LLAMA 3 be conveniently trained in a way presented below? 
The standard fine-tuning format is convenient, but I need better results, 
which probably can only be achieved with training on indices: 

`import torch
from transformers import BertTokenizer, BertForQuestionAnswering, AdamW

# Initialize tokenizer and model
model_name = "bert-base-uncased"
tokenizer = BertTokenizer.from_pretrained(model_name)
model = BertForQuestionAnswering.from_pretrained(model_name)

# Example question and context
input_text = "What is the capital of France?"
context = "France's capital is Paris, which is known for its art, culture, and history."
inputs = tokenizer(input_text, context, return_tensors="pt")

# Example start and end positions for training (indices of the tokens in the sequence)
start_positions = torch.tensor([11])  # Index of the start token 'Paris'
end_positions = torch.tensor([11])    # Index of the end token 'Paris'

# Ensure gradients can flow by not using .item()
# Forward pass to get logits and compute loss
outputs = model(**inputs, start_positions=start_positions, end_positions=end_positions)
loss = outputs.loss
start_logits = outputs.start_logits
end_logits = outputs.end_logits

# Perform a backward pass to compute gradients
loss.backward()

# Update model parameters using an optimizer
optimizer = AdamW(model.parameters(), lr=5e-5)
optimizer.step()

# Print the loss
print(f"Loss: {loss.item()}")
`

**Comment by teticio** [comment]
Can you explain what you are trying to do exactly? I'm not sure this fits into the use case I have presented. Clearly, if you want better results on SQuAD, the current SOTA are encoder only models and ensembles.

**Comment by wiiiktor** [comment]
SQuAD is not my goal, I want to create a different application, I only need to train the LLM to find position within a given text, like below: 

start_positions = torch.tensor([11]) # Index of the start token 'Paris is cool'
end_positions = torch.tensor([13]) # Index of the end token 'Paris is cool'

So, for every already text token I need some probability assesment that this should be the START and/or the END text snippet position. In the past some models (BERT) were able to find token(s) position very well and I just want to do the same, only for some other application (SQuAD would be just a useful dataset, for preliminary training).

**Comment by teticio** [comment]
OK, got it. I believe that could work thanks to the positional embeddings. Maybe you could test this with a dataset with prompts such as

Find the start and end token positions corresponding to "<<insert phrase here>>" in the following text:

....

Return your answer in JSON as follows:

{
  "start pos": ... # integer
  "end pos": ... # integer
}

and see whether this works well or not. Then pass on to the next stage of question answering.
